ddtrace/internal/datadog/profiling/dd_wrapper/src/profile.cpp

```diff
@@ -0,0 +1,228 @@
+        val_idx.wall_time = get_value_idx("wall-time", "nanoseconds");
+        val_idx.wall_count = get_value_idx("wall-samples", "count");
+    }
+    if (type_mask & SampleType::Exception) {
+        val_idx.exception_count = get_value_idx("exception-samples", "count");
+    }
+    if (type_mask & SampleType::LockAcquire) {
+        val_idx.lock_acquire_time = get_value_idx("lock-acquire-wait", "nanoseconds");
+        val_idx.lock_acquire_count = get_value_idx("lock-acquire", "count");
+    }
+    if (type_mask & SampleType::LockRelease) {
+        val_idx.lock_release_time = get_value_idx("lock-release-hold", "nanoseconds");
+        val_idx.lock_release_count = get_value_idx("lock-release", "count");
+    }
+    if (type_mask & SampleType::Allocation) {
+        val_idx.alloc_space = get_value_idx("alloc-space", "bytes");
+        val_idx.alloc_count = get_value_idx("alloc-samples", "count");
+    }
+    if (type_mask & SampleType::Heap) {
+        val_idx.heap_space = get_value_idx("heap-space", "bytes");
+    }
+
+    // Whatever the first sampler happens to be is the default "period" for the profile
+    // The value of 1 is a pointless default.
+    if (!samplers.empty()) {
+        default_period = { .type_ = samplers[0], .value = 1 };
+    }
+}
+
+size_t
+Profile::get_sample_type_length()
+{
+    return samplers.size();
+}
+
+ddog_prof_Profile&
+Profile::get_current_profile()
+{
+    return cur_profile;
+}
+
+unsigned int
+Profile::get_max_nframes() const
+{
+    return max_nframes;
+}
+
+SampleType
+Profile::get_type_mask()
+{
+    return type_mask;
+}
+
+void
+Profile::one_time_init(SampleType type, unsigned int _max_nframes)
+{
+    // In contemporary dd-trace-py, it is expected that the initialization path is in
+    // a single thread, and done only once.
+    // However, it doesn't cost us much to keep this initialization tight.
+    if (!first_time.load()) {
+        return;
+    }
+
+    // Threads need to serialize at this point
+    const std::lock_guard<std::mutex> lock(profile_mtx);
+
+    // nframes
+    max_nframes = _max_nframes;
+
+    // Set the type mask
+    const unsigned int mask_as_int = type & SampleType::All;
+    type_mask = static_cast<SampleType>(mask_as_int);
+
+    // Setup the samplers
+    setup_samplers();
+    current_pid.store(getpid());
+
+    // We need to initialize the profiles
+    const ddog_prof_Slice_ValueType sample_types = { .ptr = samplers.data(), .len = samplers.size() };
+    if (!make_profile(sample_types, &default_period, cur_profile)) {
+        std::cerr << "Error initializing top half of profile storage" << std::endl;
+        return;
+    }
+    if (!make_profile(sample_types, &default_period, last_profile)) {
+        std::cerr << "Error initializing bottom half of profile storage" << std::endl;
+        return;
+    }
+
+    // We're done. Don't do this again.
+    first_time.store(false);
+}
+
+std::string_view
+Profile::insert_or_get(std::string_view sv)
```

Comment:
When we sample with echion, we have stable strings at least for the lifetime of that sample ? 
Can some of the responsibility on string hashing be pushed to libdatadog ?
Meaning, we just use string views on the echion strings, then we let libdatadog intern things ?

Reply:
obviously this has to work beyond echion.
Though the same can be true for samples ? ++ref on that python string, --ref when we push the sample to libdatadog ?

Reply:
I think both of the ideas are good and worth pursuing.  I think we can easily solve this in two ways
1. For any interface in libdatadog which pushes a reference (pointer) type, provide a variant which immediately stores the value of the reference (probably the only type we care about is strings)
2. in this code, allow the user to send up a Python string.  Then instead of a string table, all we need to do is maintain a "dequeue" (or even a fixed array) of referenced strings to decrement at the end.  I need to think about this a bit harder though.

Echion maintains its own cache which actually makes 1 irrelevant probably.